text/0009-BundleTracker-integration.md

```diff
@@ -0,0 +1,457 @@
+<td><a href="http://docs.osgi.org/specification/osgi.core/7.0.0/util.tracker.html#d0e52020">OSGi Spec</a></td>
+<td>Must have</td>
+</tr>
+<tr class="even">
+<td>R6_Quality</td>
+<td>The solution should be robust and bug-free. This is doubly crucial as Declarative Services would then depend on the BundleTracker.</td>
+<td>PP1</td>
+<td>Must have</td>
+</tr>
+<tr class="odd">
+<td>R7_Efficiency</td>
+<td>The solution should not slow down the CppMicroServices core framework and its downstream clients.</td>
+<td><a href="http://docs.osgi.org/specification/osgi.core/7.0.0/util.tracker.html#d0e52020"></a></td>
+<td>Must have</td>
+</tr>
+</tbody>
+</table>
+
+## Functional design
+
+### Functional design summary
+
+The API for the `BundleTracker` closely mirrors that of the OSGi specification, mapped to an implementation in C++. Additionally, smart pointers are used for the `BundleTracker` API to simplify lifetime and ownership of certain objects. It is worth noting that this is different from the current implementation of the `ServiceTracker`, which still uses a raw pointer for the `ServiceTracker` constructor's `ServiceTrackerCustomizer` argument. In the future, the `ServiceTracker` API should be adapted to function like the new `BundleTracker` API to make the lifetime of the `ServiceTrackerCustomizer` more explicit.
+
+The `BundleTracker` is templated to allow for custom object types to be tracked, containing extra data or functionality.
+
+### Class definition
+
+The class definition for `BundleTracker` is as follows:
+
+```cpp
+
+namespace cppmicroservices {
+
+/**
+ * The BundleTracker class allows for keeping an accurate record of bundles and handling
+ * state changes, including bundles started before the tracker is opened.
+ *
+ * <p>
+ * A BundleTracker is constructed with a state mask and a BundleTrackerCustomizer object.
+ * The BundleTracker can select certain bundles to be tracked and trigger custom callbacks
+ * based on the provided BundleTrackerCustomizer or a BundleTracker subclass. Additionally,
+ * the template parameter allows for the use of custom tracked objects. Once the BundleTracker
+ * is opened, it will begin tracking all bundles that fall within the state mask.
+ *
+ * <p>
+ * Use the getBundles object to get the tracked Bundle objects, and getObject to get the customized
+ * object.
+ *
+ * <p>
+ * The BundleTracker class is thread-safe. It does not call BundleTrackerCustomizer methods while
+ * holding any locks. Customizer implementations must be thread-safe.
+ *
+ * @tparam T The type of tracked object.
+ * @remarks This class is thread safe.
+ */
+template <class T = Bundle>
+class BundleTracker : protected BundleTrackerCustomizer<T>
+{
+public:
+
+    /**
+     * Create a BundleTracker that tracks bundles through states covered by the state mask.
+     * 
+     * @param context The BundleContext from which tracking occurs.
+     * @param stateMask The bit mask which defines the bundle states to be tracked.
```

Comment:
I see that the OSGi spec defines that a state mask is used to specify which bundle states to track. Making this by hand isn't the most complicated thing but I think it is worth considering providing a `static uint32_t CreateStateMask(...)` utility function as part of the BundleTracker API. This is not mentioned in the OSGi spec but from a usability perspective, I'd suggest exploring example usage code of the BundleTracker and evaluating whether or not the "pain" of creating this mask manually justifies a helper.